CpsFlowExecution: parseScript(): log "Method Too Large" situations more readably

jimklimov opened the pull request:
A small almost-cosmetic change which makes life easier for my colleagues when they develop (and debug failures of) Jenkins scripted pipelines.

UPDATE: Grew to be not so small, so evicted into a method of its own to constrain changes to other parts of the codebase. Learned a few more tricks over time, as new MTL situations were found in the field.

jimklimov committed:
CpsFlowExecution: parseScript(): log "Method Too Large" situations more readably

bitwiseman commented:
Out of curiosity, what output does this give you and how is it helpful?

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -632,11 +632,15 @@ private CpsScript parseScript() throws IOException {
             for (Entry<String, String> e : loadedScripts.entrySet()) {
                 shell.reparse(e.getKey(), e.getValue());
             }
+        } catch (groovyjarjarasm.asm.MethodTooLargeException x) {
```
I do not think we want to refer to this type statically. Maybe by `simpleName`.

jimklimov replied:
Updated PR to `import` the class and catch as shorter `MethodTooLargeException`. Is it OK so?

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -632,11 +632,15 @@ private CpsScript parseScript() throws IOException {
                 shell.reparse(e.getKey(), e.getValue());
             }
+        } catch (groovyjarjarasm.asm.MethodTooLargeException x) {
+            LOGGER.log(Level.SEVERE, "FAILED to parse WorkflowScript (the pipeline script) due to MethodTooLargeException: " + x.toString());
```
Would be visible only to admins in the system log. To be useful, this would need to go into the build log. (Where it would be easily asserted by a test case.)

jimklimov replied:
Good point about a test case. Conjured up one that fits the current codebase (both master and this PR as it does not yet intrude into the build log).

jimklimov replied:
Git a bit lost in backtracking to how the "pretty message" can be shown in the build log, at least as really a message. It does not seem we have the `logger` context in a large part of the call stack where `parseScript()` is mentioned, so the best shot for visible logging seems to be to re-throw an exception with the desired message and this one as the linked cause.

jimklimov replied:
Updated below.

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge remote-tracking branch 'upstream/master' into logging-exMTL

jimklimov committed:
CpsFlowExecution: import groovyjarjarasm.asm.MethodTooLargeException to avoid catching by full name

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsScriptTest: add methodTooLargeExceptionFabricated() and methodTooLargeExceptionRealistic() tests

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution, CpsScriptTest: handle also MultipleCompilationErrorsException as a carrier of MethodTooLargeException; re-throw with just a compact message to appear in the build log

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -638,7 +639,7 @@ private CpsScript parseScript() throws IOException {
             }
-        } catch (groovyjarjarasm.asm.MethodTooLargeException x) {
+        } catch (MethodTooLargeException x) {
             LOGGER.log(Level.SEVERE, "FAILED to parse WorkflowScript (the pipeline script) due to MethodTooLargeException: " + x.toString());
```
```suggestion
```

as per https://github.com/jenkinsci/workflow-cps-plugin/pull/817/files/c438ebd0d6071eadd7fea8cd096dc5e6a51dde69..28cb0c3e54468cc8b8f527358d4c8d74b8fd9ace#r1408121159

jimklimov replied:
I guess this got outdated by the latest commit?

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -44,6 +44,7 @@
 import com.thoughtworks.xstream.io.HierarchicalStreamReader;
 import com.thoughtworks.xstream.io.HierarchicalStreamWriter;
 import com.thoughtworks.xstream.mapper.Mapper;
+import groovyjarjarasm.asm.MethodTooLargeException;
```
https://github.com/jenkinsci/workflow-cps-plugin/pull/817#discussion_r1408120262 still same issue

jimklimov replied:
PR updated in this direction

jimklimov committed:
CpsScriptTest.methodTooLargeExceptionRealistic(): reduce iteration count to satisfy different CI platforms

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution.parseScript(): handle MethodTooLargeException without importing the class

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution.parseScript(): MethodTooLargeException: actionable suggestions for pipeline devs

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsScriptTest.methodTooLargeExceptionRealistic(): update comments and expected matching line

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsScriptTest.methodTooLargeExceptionRealistic(): decouple maxStagesMethods (class complexity) from maxTryCatch (nesting depth)

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Merge branch 'master' into logging-exMTL

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -632,14 +634,74 @@ private CpsScript parseScript() throws IOException {
 
             for (Entry<String, String> e : loadedScripts.entrySet()) {
                 shell.reparse(e.getKey(), e.getValue());
             }
-        } catch (RuntimeException | Error x) {
+        } catch (Exception x) {
```
This loses handling of `Error`, especially `LinkageError`.

jimklimov replied:
I think I had some troubles compiling `catch (Exception | Error x)`, or just IDE warnings, can try.

jimklimov replied:
I see, it was an IDEA misconfiguration that had too old a "Language level" set to support multi-catches.

jimklimov replied:
Code updated to restore these `Throwable` descendant types.

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -632,14 +634,74 @@ private CpsScript parseScript() throws IOException {
+            msg += "; please refactor to simplify code structure and/or move logic to a Jenkins Shared Library: ";
+            msg += mtlEx.getMessage();
+
+            // Make a note in server log
+            LOGGER.log(Level.SEVERE, msg);
+
```
Please do not; this is up to the owner of the project to fix, not the server admin.

```suggestion
```

jimklimov replied:
After some consideration (and per comments below to another code block), I've restored this server log message - but at a "FINER" level so it is normally no-op. Still, can be an aid to Jenkins admins (e.g. developers who would reproduce the problem on their local instance).

jimklimov replied:
Later will push a revision which uses a dedicated logger (snatching the idea from "CpsFlowExecution.timing" sub-logger) for easier one-off enablement of just this logic.

jglick reviewed (commented):
The logic here is pretty convoluted and seems excessive when we do not really need to make things particularly pretty, we just need to offer _some_ clue beyond what there is today. Would suffice to do something along the lines of

```java
if (Functions.printThrowable(x).containsString("MethodTooLargeException")) {
    throw new IllegalArgumentException("…your hint here…", x);
} else {
    throw x;
}
```

jimklimov committed:
Update plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

Do not log to server console log (not all server admins are those who run the projects with issues).

Co-authored-by: Jesse Glick <jglick@cloudbees.com>

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -632,14 +634,71 @@ private CpsScript parseScript() throws IOException {
+            msg += "; please refactor to simplify code structure and/or move logic to a Jenkins Shared Library: ";
+            msg += mtlEx.getMessage();
+
+            if (ecCount > 1) {
+                // Not squashing with explicit MethodTooLargeException
+                // re-thrown below, in this codepath we have other errors.
+                throw new RuntimeException(msg, x);
+            } else {
+                // Do not confuse pipeline devs by a wall of text in the
+                // build console, but let the full context be found in
+                // server log with some dedication.
+                LOGGER.log(Level.FINE, mtlEx.getMessage());
+                //throw new RuntimeException(msg, mtlEx);
+                throw new RuntimeException(msg);
+            }
```
Regarding the concern of "pretty convoluted logic" - this block is largely why I did it so: groovy can return a `MultipleCompilationErrorsException` with an attached collection of one or more errors (broadly speaking), so we need to count if we *only* had the `MethodTooLargeException` one way or another here (and then print the pretty log), or also something else and then better print everything for human devs to sift through it diligently.

jimklimov commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -632,14 +634,71 @@ private CpsScript parseScript() throws IOException {
+                // re-thrown below, in this codepath we have other errors.
+                throw new RuntimeException(msg, x);
+            } else {
+                // Do not confuse pipeline devs by a wall of text in the
+                // build console, but let the full context be found in
+                // server log with some dedication.
+                LOGGER.log(Level.FINE, mtlEx.getMessage());
```
If we dropped that "severe" but short logging of just the prepared message into server console log that was just above (via earlier review bit => commit 89bbd4036f93a6ba8d3b226f6503c2cd6cad5046), should this detailed but "fine"-level log also disappear? Originally it was the detailed context for those couple of lines, that could be seen with a boost of verbosity, now it is probably just noise. (Or that dropped message could be revived at a less visible verbosity, e.g. make these "FINE" and "FINEST"?..)

jimklimov replied:
After some consideration (and per comment just above here), I've restored that server log message - but at a "FINER" level so it is normally no-op. Still, can be an aid to Jenkins admins (e.g. developers who would reproduce the problem on their local instance).

jimklimov commented:
Sorry it took a while to notice that there was finally a new review here - too many browser tabs open to keep track of such :D 

Thanks for doing it!

jimklimov commented:
Hm, got another use-case to handle: managed to get a step definition too large (in JSL as `vars/cloudBranch.groovy`)!

So the pipeline chugged along until it called that step - and that's where the job crashed, with a different exception:
````
[Pipeline] End of Pipeline
Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: a1490d3b-63d7-4555-ae55-3a0d6856aee6
org.jenkinsci.plugins.workflow.cps.CpsCompilationErrorsException: startup failed:
General error during class generation: Method too large: cloudBranch.___cps___586328 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;

groovyjarjarasm.asm.MethodTooLargeException: Method too large: cloudBranch.___cps___586328 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
	at groovyjarjarasm.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2087)
	at groovyjarjarasm.asm.ClassWriter.toByteArray(ClassWriter.java:447)
	at org.codehaus.groovy.control.CompilationUnit$17.call(CompilationUnit.java:850)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1087)
...
<no further pointers to the class in question>
...
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)

1 error

	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:309)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1107)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:624)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:602)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:579)
	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:323)
...
	at org.jenkinsci.plugins.workflow.cps.LoggingInvoker.getProperty(LoggingInvoker.java:121)
	at com.cloudbees.groovy.cps.impl.PropertyAccessBlock.rawGet(PropertyAccessBlock.java:20)
	at WorkflowScript.run(WorkflowScript:428)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2125)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2110)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2163)
	at WorkflowScript.run(WorkflowScript:385)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.PropertyishBlock$ContinuationImpl.get(PropertyishBlock.java:73)
	at com.cloudbees.groovy.cps.LValueBlock$GetAdapter.receive(LValueBlock.java:30)
	at com.cloudbees.groovy.cps.impl.PropertyishBlock$ContinuationImpl.fixName(PropertyishBlock.java:65)
...
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Finished: FAILURE
````

Note there are actually two lines `at WorkflowScript.run(WorkflowScript:NNN)`, pipeline code at line 385 starts an `...each {}` closure to iterate several repos this job processes, and the line 428 actually calls the `cloudBranch()` method.

Gotta look into the structure of `CpsCompilationErrorsException` to see how to best parse it here (and know if "method too large" was the only error, so the report can be trimmed), and also consider that not only `WorkflowScript` can have such troubles.

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
CpsFlowExecution: restore the multi-catch for "RuntimeException | Error"

Notably, the latter is a Throwable but not an Exception so e.g. LinkageError might be not handled.

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Revert "Update plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java
"

This reverts commit 89bbd4036f93a6ba8d3b226f6503c2cd6cad5046
to do emit the server log message but at hushed verbosity level.

jimklimov committed:
CpsFlowExecution: restore the write into server log, but only at FINE/FINER level - so it is usually hidden

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: comment that the remaining case is "ecCount == 1 exactly" and what this means practically

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
While refactoring that use-case, I've also got a nested hit, with pipeline and several JSL steps being okay, but a further step (evicted code) being too large:
````
Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: 32491c0c-247d-4340-a173-b778431b7d6a
org.jenkinsci.plugins.workflow.cps.CpsCompilationErrorsException: startup failed:
General error during class generation: Method too large: branchResync.___cps___636753 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;

groovyjarjarasm.asm.MethodTooLargeException: Method too large: branchResync.___cps___636753 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
	at groovyjarjarasm.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2087)
	at groovyjarjarasm.asm.ClassWriter.toByteArray(ClassWriter.java:447)
...
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)

1 error

	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:309)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1107)
...
	at com.cloudbees.groovy.cps.impl.PropertyAccessBlock.rawGet(PropertyAccessBlock.java:20)
	at cloudBranch.resync(cloudBranch.groovy:78)
	at cloudBranch.prepareFeatureFreeze(cloudBranch.groovy:220)
	at cloudBranch.processFeatureFreeze(cloudBranch.groovy:564)
	at WorkflowScript.run(WorkflowScript:428)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2125)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2110)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2163)
	at WorkflowScript.run(WorkflowScript:385)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.PropertyishBlock$ContinuationImpl.get(PropertyishBlock.java:73)
...
````

jimklimov commented:
And another pattern when the code was moved into a JSL class:
````
Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: 99b19442-dee8-46d4-8d05-05439214279a
org.jenkinsci.plugins.workflow.cps.CpsCompilationErrorsException: startup failed:
General error during class generation: Method too large: com/myprovys/ci/BranchResync.___cps___679414 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;

groovyjarjarasm.asm.MethodTooLargeException: Method too large: com/myprovys/ci/BranchResync.___cps___679414 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
	at groovyjarjarasm.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2087)
	at groovyjarjarasm.asm.ClassWriter.toByteArray(ClassWriter.java:447)
	at org.codehaus.groovy.control.CompilationUnit$17.call(CompilationUnit.java:850)
...
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)

1 error

	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:309)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1107)
...
	at com.cloudbees.groovy.cps.impl.PropertyAccessBlock.rawGet(PropertyAccessBlock.java:20)
	at WorkflowScript.run(WorkflowScript:428)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2125)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2110)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2163)
	at WorkflowScript.run(WorkflowScript:385)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.PropertyishBlock$ContinuationImpl.get(PropertyishBlock.java:73)
...
````
(here via direct call from pipeline `WorkflowScript`)

jimklimov committed:
CpsFlowExecution: recognize also CpsCompilationErrorsException for MethodTooLargeException handling

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution, CpsScriptTest: recognize also Groovy (JSL step or class) for MethodTooLargeException handling

And re-word logging (impacted class name, possible excerpts from trace).
Not only WorkflowScript (pipeline) suffers from this, see examples
in https://github.com/jenkinsci/workflow-cps-plugin/pull/817

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
Posting the updated code to recognize the newly seen exception from Method Too Large in a JSL. I'll be testing a build with this on servers I manage, to get some first-hand feedback.

A few things of note are:
* we try to detect the problematic step/class name (if not `WorkflowScript`) and make less assumption that specifically the pipeline script code is too large
* newly emitted stack trace (to be seen in build log) should include a trail through `*.groovy` source files+line numbers, if present (assuming JSL)
* logging into server log only goes as FINE(R) so should not be a problem by default (but developers figuring out the issue with a private Jenkins instance can drill into full gory details on demand)

jimklimov commented:
Heh, the code path which leads to MTL stack trace for steps/classes from JSL apparently lands into a different method's try/catch. Reworking the plugin change for the not-`WorkflowScript` case, although currently inclined to make this handling a method to just call from several places.

jimklimov committed:
CpsFlowExecution: introduce METHOD_TOO_LARGE_LOGGER for separated logging of pretty-printing certain error handling

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: annotate what we post into METHOD_TOO_LARGE_LOGGER

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
Figured it out locally, preparing refactored PR contents as a meaningful series of commits :)

jimklimov committed:
CpsFlowExecution: refactor whole reportSuspectedMethodTooLarge() magic from parseScript() into a method of its own

I first wanted it to just throw the exceptions, but IDEA was upset that
we might return from the catch block and then use the uninitialized "s"
variable. So this new method returns the entity which we clearly throw
from the catch block.

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): fix indentations after refactoring out of earlier codebase

Also avoid the final if/else to have a clear final code path

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): when returning short exception with a curated message to go into job log, chop off the stack trace leading to the reporter

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): xStr should not include x.getMessage(), a copy is already there

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): at least CpsCompilationErrorsException.getMessage() is too long with all the original stack trace

Introduce an xMsgStart with curated start of original Throwable's
message (until the first "\tat somewhere(file:line)" match).

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): describe what we see in pretty-printed overflowedClassName

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): document MTLE_CLASSNAME_PATTERN and CLASSNAME_MENTIONS_PATTERN expectations/assumptions in code

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): overflowedClassName may have slashes for class or not for step

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): support also pipeline scripts named like "Script<NUM>"

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: parseScript(): IDEA complains for "throw reportSuspectedMethodTooLarge()" as too generic

Curiously, before this PR #817 it had no qualms to "throw x"
which could be a RuntimeException or Error, and the method
itself only declares that it "throws IOException" which is
neither. Even now explicitly-cast rethrows seem acceptable :\

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsScript: getProperty(): involve CpsFlowExecution.reportSuspectedMethodTooLarge(x)

Some loaded properties are in fact calls to resolve code from a
Jenkins Shared Library (steps aka global variables, or directly
class methods if using script{} blocks and not just declarative).
First accesses to such code cause CPS, compilation and possible
failure with a MethodTooLargeException (maybe hidden in another
Throwable) which we also want pretty-printed to be actionable
for a pipeline developer.

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): wrap original log excerpts into snip-markers to separate visibly

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): refactor actionableMsg as a StringBuilder

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): refactor overflowedClassNameBreadcrumbs as a StringBuilder

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): match by *short* overflowedClassName for overflowedClassNameBreadcrumbs

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): wrap original log excerpts - only add a newline before final wrapper if not present in wrapped text

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): comment about "interaction" with ContinuationGroup.fixupStackTrace()

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): do not hijack mtlEx in the end, use a dedicated "RuntimeException rtex" object

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: reportSuspectedMethodTooLarge(): avoid extra blank line

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
Updated screenshots, now with some detail learned about MTL issues happening in the JSL also:

* Pipeline script too large:
````
...
Executing git in workdir C:\Users\klimov\.jenkins\workspace\test-MethodTooLarge@libs\3fbaa4c2a74a618783608223aa05bcd2484d251d9f183ec393764a9458686f38
java.lang.RuntimeException: FAILED to parse WorkflowScript (the pipeline script) due to MethodTooLargeException; please refactor to simplify code structure and/or move logic to a Jenkins Shared Library:
-----
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
General error during class generation: Method too large: WorkflowScript.___cps___5112 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
groovyjarjarasm.asm.MethodTooLargeException: Method too large: WorkflowScript.___cps___5112 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
-----

Complete details can be seen in server log at FINE/FINER level (Jenkins admin access for org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.MethodTooLargeLogging is required)
Finished: FAILURE
````

* JSL step:
````
...
[Pipeline] End of Pipeline
Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: e5f57f7d-1081-49cf-a579-0644086eead0
java.lang.RuntimeException: FAILED to parse presumed JSL step 'cloudBranch' due to MethodTooLargeException; please refactor to simplify code structure:
-----
org.jenkinsci.plugins.workflow.cps.CpsCompilationErrorsException: startup failed:
General error during class generation: Method too large: cloudBranch.___cps___1028 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
groovyjarjarasm.asm.MethodTooLargeException: Method too large: cloudBranch.___cps___1028 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
-----

Complete details can be seen in server log at FINE/FINER level (Jenkins admin access for org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.MethodTooLargeLogging is required)
Finished: FAILURE
````

* JSL class:
````
...
[Pipeline] End of Pipeline
Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: 9338d5ab-bd20-4f09-b81a-2d13b2eaf911
java.lang.RuntimeException: FAILED to parse presumed JSL class 'com/myprovys/ci/BranchResync' due to MethodTooLargeException; please refactor to simplify code structure:
-----
org.jenkinsci.plugins.workflow.cps.CpsCompilationErrorsException: startup failed:
General error during class generation: Method too large: com/myprovys/ci/BranchResync.___cps___2123 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
groovyjarjarasm.asm.MethodTooLargeException: Method too large: com/myprovys/ci/BranchResync.___cps___2123 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
-----

Complete details can be seen in server log at FINE/FINER level (Jenkins admin access for org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.MethodTooLargeLogging is required)
Finished: FAILURE
````

Notably, code was made to trace the messages I've seen (and posted earlier above) with "bread-crumbs" to see the code path through the pipeline script and groovy (JSL) files and line numbers (for steps/classes with code too large), only to discover this is slapped on later via `ContinuationGroup.fixupStackTrace()` called from PropertyishBlock.ContinuationImpl.get()` :\

Probably some refactoring can be due to take advantage of that bit. At least, the groovy-cps code is now in the same plugin.

jimklimov committed:
CpsScriptTest: methodTooLargeExceptionRealistic(): expect updated wording for logged message

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge branch 'master' into logging-exMTL

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -642,21 +648,324 @@ Invoker createInvoker() {
+        }
+        actionableMsg.append("; please refactor to simplify code structure");
+        if (overflowedClassNameReport.contains("WorkflowScript") || CLASSNAME_SCRIPTNUM_PATTERN.matcher(overflowedClassName).find())
+            actionableMsg.append(" and/or move logic to a Jenkins Shared Library");
```
BTW the terminology in https://www.jenkins.io/doc/book/pipeline/shared-libraries/ is unfortunate. By definition a “library” is “shared”; why else would you create a library? The [plugin itself](https://github.com/jenkinsci/pipeline-groovy-lib-plugin/) uses the term _Pipeline Groovy library_.

jimklimov replied:
Updated to mention both terminologies, hopefully that makes it clearer to both populations of readers.

OTOH, I often see (and use) the acronym JSL though, but never saw a PGL...

jimklimov replied:
Although I would say that code-sharing is one aspect. Code offloading (into a separate compilable unit) is another :)
Some organizations might want the looks of a declarative pipeline for some job with their custom logic behind it, so the logic is in a library (trusted, un-sandboxed and all) while the pipeline seems to call a few steps with a clean interface. But it may be the only consumer (the not-quite-sharing aspect of this).

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -466,6 +470,8 @@ Timing time(TimingKind kind) {
 
     static final Logger TIMING_LOGGER = Logger.getLogger(CpsFlowExecution.class.getName() + ".timing");
 
+    static final Logger METHOD_TOO_LARGE_LOGGER = Logger.getLogger(CpsFlowExecution.class.getName() + ".MethodTooLargeLogging");
```
Is there some reason not to simply use https://github.com/jenkinsci/workflow-cps-plugin/blob/7fd3dd259bf2bb92da8916bf5bfde98733ad23bc/plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java#L2200 ?

jimklimov replied:
Something about keeping the possibly large messages out of sight unless truly wanted in troubleshooting.

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -642,21 +648,324 @@ Invoker createInvoker() {
+        //rtex.addSuppressed(mtlEx);
+
+        return rtex;
+    }
```
I am just going to note the irony of a 285-line method here. :grin:

jimklimov replied:
Noted :)

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -642,21 +648,324 @@ Invoker createInvoker() {
+        Throwable mtlEx = null;
+        int ecCount = 0;
+        String xStr = Functions.printThrowable(x); // includes x.getMessage() contents
+        final Pattern LINE_SEP_PATTERN = Pattern.compile("\\R");
```
I had to look this one up.

> Any Unicode linebreak sequence, is equivalent to `\u000D\u000A|[\u000A\u000B\u000C\u000D\u0085\u2028\u2029]`

AFAICT
```suggestion
        final Pattern LINE_SEP_PATTERN = Pattern.compile("\r?\n");
```
suffices.

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -642,21 +648,324 @@ Invoker createInvoker() {
+                .append(METHOD_TOO_LARGE_LOGGER.getName())
+                .append(" is required)");
+
+        //return new RuntimeException(actionableMsg.toString(), mtlEx);
```
Do not leave commented-out code in a final PR (unless there is a real possibility it should be resurrected some day soonish and a preceding comment explains those circumstances).

jglick commented on plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java:
```diff
@@ -642,21 +648,324 @@ Invoker createInvoker() {
+        } else if (x instanceof CpsCompilationErrorsException) {
+            // Defined in this plugin, to clone a message and stack trace
+            // from a MultipleCompilationErrorsException and be serializable.
+            // Grep it as text for "MethodTooLargeException" and "1 error"
```
Would it not make more sense to just enrich `CpsCompilationErrorsException` with whatever information you need? AFAICT all you are looking for ultimately is the `MethodTooLargeException.message`. So, we could just keep that in a field?

jimklimov replied:
Interesting idea actually. At the time this was written, I did not realize such a possibility on one hand, and probably wanted the changes to not metastacize all over the codebase on another.

jglick commented on plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScriptTest.java:
```diff
@@ -92,4 +92,124 @@ public void evaluateShallSandbox() throws Exception {
         r.assertLogContains("Scripts not permitted to use method groovy.lang.Script run java.io.File java.lang.String[]", b);
     }
 
+    @Test public void methodTooLargeExceptionFabricated() throws Exception {
+        // Fabricate a MethodTooLargeException which "normally" happens when evaluated
```
Is there a reason to keep this test case, given the “realistic” test below?

jimklimov replied:
I think yes, this fabricated test is about coded reaction to the particular exception, which may be or not be the one thrown in realistic test below (if CPS VM does at all break such way there and with that amount of threads).

jglick commented on plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScriptTest.java:
```diff
@@ -92,4 +92,124 @@ public void evaluateShallSandbox() throws Exception {
+        StringBuffer sbStages = new StringBuffer();
+
+        // Limits to the "max":
+        // * java.lang.StackOverflowError varies per JDK platform
+        //   (CI on Linux was unhappy with 255, on Windows with 1023)
```
You _can_ set a per-thread stack size, but unfortunately the unit of measurement is unspecified and presumably specific to JVM implementation details, so this is probably not that helpful.

jglick replied:
(Why would you would be getting a JVM-level SOE, anyway? This is all dealing with the CPS VM that does not use the native stack.)

jimklimov replied:
Saw some code in this plugin throwing it (for "excessively nested" cases at least), so thought it is worth keeping in mind.

jglick reviewed (approved):
OK I think, though I can barely follow most of the code here. @dwnusbaum did you expect to review?

dwnusbaum commented:
I have no plans to review this. My impression from a very brief look at the code was that it seems quite complicated, but I am not blocking it. IMO creating a page like https://www.jenkins.io/doc/book/pipeline/cps-method-mismatches/ explaining the situation and recommendations in detail seems like it could help, and then perhaps this code could be simplified to only check the stack trace for the relevant class types and add a suppressed exception that links to that page, accepting potential false positives and not adding the additional context currently added by this PR (but maybe that context is critically useful, IDK).

jglick commented:
> only check the stack trace for the relevant class types and add a suppressed exception

https://github.com/jenkinsci/workflow-cps-plugin/pull/817#pullrequestreview-2029698582 more or less. IIUC this PR offers the same actionable information as was already there, but formatted differently?

> links to that page

Actually a redirect URL like in https://github.com/jenkins-infra/jenkins.io/pull/2290.

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov commented:
I'm not sure how the simple redirect to a wiki page would be better?

Do we not-print a wall of text and redirect to a wiki saying "sometimes the script is too big, guess which one of yours that is"?

Or do we print the wall of text and redirect to a wiki that says same and "here's how to look for a needle in the haystack"?

Currently the changeset pinpoints which pipeline, step or JSL class has upset the parser (finds those needles in the haystack for us), so I am not sure how the other options are better. Surely a Wiki link can be posted "in addition" to this, but "instead" - I am not so excited about that path (even not due to author bias - just due to loss of usability improvement we use for over a year).

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge remote-tracking branch 'upstream/master' as of 2025-10-04 into logging-exMTL

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution, CpsScriptTest: mvn spotless:apply

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution: rephrase Jenkins Shared Library/JSL (doc book term) to mention Pipeline Groovy library/PGL (plugin term)

@jglick> BTW the terminology in https://www.jenkins.io/doc/book/pipeline/shared-libraries/
is unfortunate. By definition a “library” is “shared”; why else would you create a library?
The plugin itself uses the term Pipeline Groovy library.

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CpsFlowExecution:: simplify end-of-line regex

Apply suggestion from @jglick

Co-authored-by: Jesse Glick <jglick@cloudbees.com>

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov committed:
Merge branch 'master' into logging-exMTL

jimklimov commented:
Another codepath now led to the exception so it was not parsed. Re-appreciated the niceness of:
```
...
22:30:18   > git.exe rev-list --no-walk 9c3718a88fd3ccae07b4e70b04e29a3896f18c75 # timeout=10
22:30:26  [Pipeline] Start of Pipeline
22:30:29  [Pipeline] End of Pipeline
22:30:29  Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: 557646c8-16bf-46ba-ba70-6217afb6b2c7
22:30:29  java.lang.RuntimeException: FAILED to parse presumed PGL (JSL) step 'dynamatrixPipeline' due to MethodTooLargeException; please refactor to simplify code structure:
22:30:29  -----
22:30:29  org.jenkinsci.plugins.workflow.cps.CpsCompilationErrorsException: startup failed:
22:30:29  General error during class generation: Method too large: dynamatrixPipeline.___cps___120 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
22:30:29  groovyjarjarasm.asm.MethodTooLargeException: Method too large: dynamatrixPipeline.___cps___120 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
22:30:29  -----
22:30:29  Complete details can be seen in server log at FINE/FINER level (Jenkins admin access for org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.MethodTooLargeLogging is required)
22:30:29  Finished: FAILURE
```
(maybe should have had a bit of `...	at WorkflowScript.run(WorkflowScript:1401)` too), against the monstrous:
```
17:55:47   > git.exe rev-list --no-walk 9c3718a88fd3ccae07b4e70b04e29a3896f18c75 # timeout=10
17:55:48  [Pipeline] Start of Pipeline
17:56:22  [Pipeline] End of Pipeline
17:56:22  Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: a788ad05-d11b-42de-9bc9-3cf080ce64a0
17:56:22  org.jenkinsci.plugins.workflow.cps.CpsCompilationErrorsException: startup failed:
17:56:22  General error during class generation: Method too large: dynamatrixPipeline.___cps___628 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
17:56:22  
17:56:22  groovyjarjarasm.asm.MethodTooLargeException: Method too large: dynamatrixPipeline.___cps___628 ()Lcom/cloudbees/groovy/cps/impl/CpsFunction;
17:56:22  	at groovyjarjarasm.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2087)
17:56:22  	at groovyjarjarasm.asm.ClassWriter.toByteArray(ClassWriter.java:447)
17:56:22  	at org.codehaus.groovy.control.CompilationUnit$17.call(CompilationUnit.java:850)
17:56:22  	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1087)
17:56:22  	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:624)
17:56:22  	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:602)
17:56:22  	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:579)
17:56:22  	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:323)
17:56:22  	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:293)
17:56:22  	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:279)
17:56:22  	at groovy.lang.GroovyClassLoader.recompile(GroovyClassLoader.java:786)
17:56:22  	at groovy.lang.GroovyClassLoader.loadClass(GroovyClassLoader.java:743)
17:56:22  	at groovy.lang.GroovyClassLoader.loadClass(GroovyClassLoader.java:812)
17:56:22  	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:544)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsGroovyShell$TimingLoader.loadClass(CpsGroovyShell.java:231)
17:56:22  	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:544)
17:56:22  	at groovy.lang.GroovyClassLoader.loadClass(GroovyClassLoader.java:702)
17:56:22  	at groovy.lang.GroovyClassLoader.loadClass(GroovyClassLoader.java:812)
17:56:22  	at groovy.lang.GroovyClassLoader.loadClass(GroovyClassLoader.java:800)
17:56:22  	at PluginClassLoader for pipeline-groovy-lib//org.jenkinsci.plugins.workflow.cps.global.UserDefinedGlobalVariable.getValue(UserDefinedGlobalVariable.java:52)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsScript.invokeMethod(CpsScript.java:115)
17:56:22  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
17:56:22  	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
17:56:22  	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
17:56:22  	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
17:56:22  	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1225)
17:56:22  	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
17:56:22  	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:41)
17:56:22  	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
17:56:22  	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
17:56:22  	at PluginClassLoader for script-security//org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:180)
17:56:22  	at PluginClassLoader for script-security//org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:23)
17:56:22  	at PluginClassLoader for script-security//org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:163)
17:56:22  	at PluginClassLoader for script-security//org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:148)
17:56:22  	at PluginClassLoader for script-security//org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:178)
17:56:22  	at PluginClassLoader for script-security//org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:182)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:17)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.LoggingInvoker.methodCall(LoggingInvoker.java:124)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:107)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:118)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:87)
17:56:22  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
17:56:22  	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:71)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.LocalVariableBlock$LocalVariable.get(LocalVariableBlock.java:39)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.LValueBlock$GetAdapter.receive(LValueBlock.java:30)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.LocalVariableBlock.evalLValue(LocalVariableBlock.java:28)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.LValueBlock$BlockImpl.eval(LValueBlock.java:54)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.LValueBlock.eval(LValueBlock.java:16)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.Next.step(Next.java:84)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:142)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:17)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:48)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:188)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:464)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:372)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:302)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService.lambda$wrap$4(CpsVmExecutorService.java:143)
17:56:22  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
17:56:22  	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:139)
17:56:22  	at jenkins.util.ContextResettingExecutorService.lambda$wrap$0(ContextResettingExecutorService.java:26)
17:56:22  	at jenkins.security.ImpersonatingExecutorService.lambda$wrap$0(ImpersonatingExecutorService.java:66)
17:56:22  	at jenkins.util.ErrorLoggingExecutorService.lambda$wrap$0(ErrorLoggingExecutorService.java:51)
17:56:22  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
17:56:22  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
17:56:22  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
17:56:22  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$1.call(CpsVmExecutorService.java:53)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$1.call(CpsVmExecutorService.java:50)
17:56:22  	at org.codehaus.groovy.runtime.GroovyCategorySupport$ThreadCategoryInfo.use(GroovyCategorySupport.java:136)
17:56:22  	at org.codehaus.groovy.runtime.GroovyCategorySupport.use(GroovyCategorySupport.java:275)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService.lambda$categoryThreadFactory$0(CpsVmExecutorService.java:50)
17:56:22  	at java.base/java.lang.Thread.run(Thread.java:1474)
17:56:22  
17:56:22  1 error
17:56:22  
17:56:22  	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:309)
17:56:22  	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1107)
17:56:22  	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:624)
17:56:22  	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:602)
17:56:22  	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:579)
17:56:22  	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:323)
17:56:22  	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:293)
17:56:22  	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:279)
17:56:22  	at groovy.lang.GroovyClassLoader.recompile(GroovyClassLoader.java:786)
17:56:22  	at groovy.lang.GroovyClassLoader.loadClass(GroovyClassLoader.java:743)
17:56:22  	at groovy.lang.GroovyClassLoader.loadClass(GroovyClassLoader.java:812)
17:56:22  	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:544)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsGroovyShell$TimingLoader.loadClass(CpsGroovyShell.java:231)
17:56:22  	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:544)
17:56:22  	at groovy.lang.GroovyClassLoader.loadClass(GroovyClassLoader.java:702)
17:56:22  	at groovy.lang.GroovyClassLoader.loadClass(GroovyClassLoader.java:812)
17:56:22  	at groovy.lang.GroovyClassLoader.loadClass(GroovyClassLoader.java:800)
17:56:22  	at PluginClassLoader for pipeline-groovy-lib//org.jenkinsci.plugins.workflow.cps.global.UserDefinedGlobalVariable.getValue(UserDefinedGlobalVariable.java:52)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsScript.invokeMethod(CpsScript.java:115)
17:56:22  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
17:56:22  	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
17:56:22  	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
17:56:22  	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
17:56:22  	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1225)
17:56:22  	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
17:56:22  	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:41)
17:56:22  	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
17:56:22  	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
17:56:22  	at PluginClassLoader for script-security//org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:180)
17:56:22  	at PluginClassLoader for script-security//org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:23)
17:56:22  	at PluginClassLoader for script-security//org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:163)
17:56:22  	at PluginClassLoader for script-security//org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:148)
17:56:22  	at PluginClassLoader for script-security//org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:178)
17:56:22  	at PluginClassLoader for script-security//org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:182)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:17)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.LoggingInvoker.methodCall(LoggingInvoker.java:124)
17:56:22  	at WorkflowScript.run(WorkflowScript:1401)
17:56:22  	at ___cps.transform___(Native Method)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:107)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:118)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:87)
17:56:22  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
17:56:22  	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:71)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.LocalVariableBlock$LocalVariable.get(LocalVariableBlock.java:39)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.LValueBlock$GetAdapter.receive(LValueBlock.java:30)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.impl.LocalVariableBlock.evalLValue(LocalVariableBlock.java:28)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.LValueBlock$BlockImpl.eval(LValueBlock.java:54)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.LValueBlock.eval(LValueBlock.java:16)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.Next.step(Next.java:84)
17:56:22  	at PluginClassLoader for workflow-cps//com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:142)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:17)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:48)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:188)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:464)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:372)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:302)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService.lambda$wrap$4(CpsVmExecutorService.java:143)
17:56:22  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
17:56:22  	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:139)
17:56:22  	at jenkins.util.ContextResettingExecutorService.lambda$wrap$0(ContextResettingExecutorService.java:26)
17:56:22  	at jenkins.security.ImpersonatingExecutorService.lambda$wrap$0(ImpersonatingExecutorService.java:66)
17:56:22  	at jenkins.util.ErrorLoggingExecutorService.lambda$wrap$0(ErrorLoggingExecutorService.java:51)
17:56:22  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
17:56:22  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
17:56:22  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
17:56:22  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$1.call(CpsVmExecutorService.java:53)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$1.call(CpsVmExecutorService.java:50)
17:56:22  	at org.codehaus.groovy.runtime.GroovyCategorySupport$ThreadCategoryInfo.use(GroovyCategorySupport.java:136)
17:56:22  	at org.codehaus.groovy.runtime.GroovyCategorySupport.use(GroovyCategorySupport.java:275)
17:56:22  	at PluginClassLoader for workflow-cps//org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService.lambda$categoryThreadFactory$0(CpsVmExecutorService.java:50)
17:56:22  	at java.base/java.lang.Thread.run(Thread.java:1474)
17:56:22  Finished: FAILURE
```

When can this be merged? ;)
